### PR TITLE
Alerting: Expressions pipeline redesign

### LINKED ISSUE #46176: Alerting: redesign expression pipeline
The current expressions pipeline is taking up a large chunk of screen real-estate – we should rethink its design to

1. [ ] Be more succinct and take up less space
2. [x] Improve the experience of hooking up expressions (the user shouldn't have to update ref IDs all the time)
https://github.com/grafana/grafana/pull/51810
3. [ ] Allow setting the alert condition from an expression itself

### PR DESCRIPTION
**Pull request by gillesdemey**
image

Somewhat long overdue and a rather sizable PR, but this introduces a new redesigned expressions pipeline editor that combines both the expressions editor and the preview in to a single view.

**Notable changes in the PR:**

* 🎨 New design 
* 🗑️ Preview has been merged in to the expressions pipeline (this means it's also been removed for Grafana managed rule) 
* 🌊 Data flow has been changed substantially since the `QueryEditor` no longer manages the expressions but only the data queries
* 🧮 Some expressions (mostly Math expression) have an updated design to accommodate a smaller card design
* ✨ The "Reduce + Math" expression combo is now the default instead of the "Classic Condition" 

**Which issue(s) this PR fixes**:
 
Fixes #46176

**Special notes for your reviewer**:

Since this PR is rather large I invite the reviewers to check out the branch locally, play around with it and hunt for bugs and we can schedule a mob programming session to walk through some of the most significant changes, probably focusing mostly on the new data flow and reducers.

**Some of the next steps I'd like to take are:**

* Adding a "Threshold" expression that acts much like the "Classic Condition" but needs to be combined with a "reduce" expression and would support multi-dimensional alerts. This will be the new default pipeline (reduce + threshold).
* I still have to fix and run tests, this might take a while to get done – in the mean time CI will probably fail 👷

### CODE REVIEW

File: public/app/features/expressions/types.ts [hunk 1]
```diff
@@ -10,10 +10,28 @@ export enum ExpressionQueryType {
 }
 
 export const gelTypes: Array<SelectableValue<ExpressionQueryType>> = [
-  { value: ExpressionQueryType.math, label: 'Math' },
-  { value: ExpressionQueryType.reduce, label: 'Reduce' },
-  { value: ExpressionQueryType.resample, label: 'Resample' },
-  { value: ExpressionQueryType.classic, label: 'Classic condition' },
+  {
```
Comment (gillesdemey) [inline public/app/features/expressions/types.ts:13]: 💬 I've just added some descriptions to these expression types here

File: public/app/features/expressions/components/Resample.tsx [hunk 2]
```diff
@@ -8,11 +8,11 @@ import { downsamplingTypes, ExpressionQuery, upsamplingTypes } from '../types';
 interface Props {
   refIds: Array<SelectableValue<string>>;
   query: ExpressionQuery;
-  labelWidth: number;
+  labelWidth?: number | 'auto';
```
Comment (gillesdemey) [inline public/app/features/expressions/components/Resample.tsx:11]: 💬 The downstream component supports `auto` as `labelWidth` but this not supported in the typing of this component.

File: public/app/features/expressions/components/Math.tsx [hunk 3]
```diff
@@ -36,93 +34,97 @@ export const Math: FC<Props> = ({ labelWidth, onChange, query, onRunQuery }) =>
   return (
     <Stack direction="row">
       <InlineField
-        label="Expression"
```
Comment (gillesdemey) [inline public/app/features/expressions/components/Math.tsx:39]: 💬 The changes in this file are mostly to refactor the `Math` expression help text, to support a narrower card view I moved the info to a hover card that is shown when you hover over the "Expression (?)" label. This also works nicer with the panel view (where we also support expressions).

File: public/app/features/expressions/components/Condition.tsx [hunk 4]
```diff
@@ -69,65 +69,70 @@ export const Condition: FC<Props> = ({ condition, index, onChange, onRemoveCondi
     condition.evaluator.type === EvalFunction.IsWithinRange || condition.evaluator.type === EvalFunction.IsOutsideRange;
 
   return (
-    <InlineFieldRow>
```
Comment (gillesdemey) [inline public/app/features/expressions/components/Condition.tsx:72]: 💬 I've mostly rearranged some fields in this `Classic Condition` expression to support a narrow card view, the inputs are no longer all one one line but split across multiple lines.

File: public/app/features/alerting/unified/utils/rule-form.ts [hunk 5]
```diff
@@ -312,8 +354,16 @@ export const panelToRuleFormValues = async (
     return undefined;
   }
 
+  // TODO make this work with correct refids
   if (!queries.find((query) => query.datasourceUid === ExpressionDatasourceUID)) {
-    queries.push(getDefaultExpression(getNextRefIdChar(queries.map((query) => query.model))));
+    const [reduceExpression, _mathExpression] = getDefaultExpressions(getNextRefIdChar(queries), '-');
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/utils/rule-form.ts:359]: 💬 This is a bit weird, but the `getDefaultExpressions()` function wants the `refIds` and I don't have them yet, so I'm adding each expression one at a time to get the correct `refId`.

File: public/app/features/alerting/unified/utils/rule-form.ts [hunk 6]
```diff
@@ -188,15 +188,18 @@ export const getDefaultQueries = (): AlertQuery[] => {
...
     },
-    getDefaultExpression('B'),
+    ...getDefaultExpressions('B', 'C'),
   ];
 };
 
-const getDefaultExpression = (refId: string): AlertQuery => {
-  const model: ExpressionQuery = {
-    refId,
+const getDefaultExpressions = (...refIds: [string, string]): AlertQuery[] => {
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/utils/rule-form.ts:195]: 💬 I've changed the default to a chain of "reduce + math" expression for new Grafana managed alerts.

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+export const setDataQueries = createAction<AlertQuery[]>('setDataQueries');
+
+export const addNewExpression = createAction('addNewExpression');
+export const removeExpression = createAction<string>('removeExpression');
+export const updateExpression = createAction<ExpressionQuery>('updateExpression');
+export const updateExpressionRefId = createAction<{ oldRefId: string; newRefId: string }>('updateExpressionRefId');
+export const rewireExpressions = createAction<{ oldRefId: string; newRefId: string }>('rewireExpressions');
+export const updateExpressionType = createAction<{ refId: string; type: ExpressionQueryType }>('updateExpressionType');
+
+export const queriesAndExpressionsReducer = createReducer(initialState, (builder) => {
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts:35]: 💬 I'm using `createReducer` here from `@redux/toolkit`, which works with React's `useReducer` but comes with support for `Immer` out-of-the-box when we add our reduce functions.

Each reduce function receives `state` which is a "draft" that you can mutate, Immer will take care of computing the correct delta and apply it immutably.

https://redux-toolkit.js.org/usage/immer-reducers

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+  updateExpression,
+  updateExpressionRefId,
+  updateExpressionType,
+} from './reducer';
+
+interface Props {
+  editingExistingRule: boolean;
+}
+
+export const QueryAndExpressionsStep: FC<Props> = ({ editingExistingRule }) => {
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx:37]: 💬 This component has gotten quite large but I'm not sure if adding additional abstractions is a good idea at this time, we might enter the realm of diminishing returns by adding too many abstractions and passing everything as props or function args.

Open to feedback on this one for sure!

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+  const dataQueries = useMemo(() => {
+    return queries.filter((query) => !isExpressionQuery(query.model));
+  }, [queries]);
+
+  const emptyQueries = queries.length === 0;
+
+  const onUpdateRefId = useCallback(
+    (oldRefId: string, newRefId: string) => {
+      const newRefIdExists = refIdExists(queries, newRefId);
+      // TODO we should set an error and explain what went wrong instead of just refusing to update
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx:104]: 💬 I think this is pretty much the only `TODO` I haven't figured out yet, any ideas / thoughts?

File: public/app/features/alerting/unified/components/rule-editor/QueryRows.tsx [hunk 10]
```diff
@@ -162,11 +148,8 @@ export class QueryRows extends PureComponent<Props, State> {
     onQueriesChange(update);
   };
 
-  onDuplicateQuery = (query: DataQuery, source: AlertQuery): void => {
-    this.props.onDuplicateQuery({
-      ...source,
-      model: query,
-    });
+  onDuplicateQuery = (query: DataQuery): void => {
+    throw new Error('not yet implemented');
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rule-editor/QueryRows.tsx:152]: 💬 Oops, gotta add this one back in

File: public/app/features/alerting/unified/components/rule-editor/QueryEditor.tsx [hunk 11]
```diff
@@ -1,224 +1,40 @@
 import { css } from '@emotion/css';
-import React, { PureComponent } from 'react';
+import React, { FC } from 'react';
 
-import {
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rule-editor/QueryEditor.tsx:4]: 💬 This component has been cleaned up significantly as it was previously responsible for managing both data queries _and_ expression queries.

This logic has been moved to both the `QueriesAndExpressionsStep` and a separate `reducer.ts` to manage the data structures.

File: public/app/features/alerting/unified/components/rule-editor/ExpressionsEditor.tsx [hunk 12]
```diff
@@ -0,0 +1,56 @@
+import React, { FC, useMemo } from 'react';
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rule-editor/ExpressionsEditor.tsx:1]: 💬 This one is mostly a wrapper around the `Expression` component and renders one for each expression query we have.

File: public/app/features/alerting/unified/components/rule-editor/ConditionField.tsx [hunk 13]
```diff
@@ -1,95 +0,0 @@
-import { css } from '@emotion/css';
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/rule-editor/ConditionField.tsx:1]: 💬 We no longer need this one as conditions are now set from the expression of data query directly.

File: public/app/features/alerting/unified/components/expressions/Expression.tsx [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+        <div className={styles.expression.resultValue}>{value}</div>
+        {showFiring && <AlertStateTag state={PromAlertingRuleState.Firing} size="sm" />}
+        {showNormal && <AlertStateTag state={PromAlertingRuleState.Inactive} size="sm" />}
+      </Stack>
+    </div>
+  );
+};
+
+// TODO: move this to design system or util file?
+const Spacer = () => <span style={{ flex: 1 }}></span>;
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/expressions/Expression.tsx:230]: 💬 Any thought on whether we should move it somewhere? It's currently just a `div` I've used to push flex elements away from each other.

File: public/app/features/alerting/unified/components/expressions/Expression.tsx [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+  queries: AlertQuery[];
+  query: ExpressionQuery;
+  onSetCondition: (refId: string) => void;
+  onUpdateRefId: (oldRefId: string, newRefId: string) => void;
+  onRemoveExpression: (refId: string) => void;
+  onUpdateExpressionType: (refId: string, type: ExpressionQueryType) => void;
+  onChangeQuery: (query: ExpressionQuery) => void;
+}
+
+export const Expression: FC<ExpressionProps> = ({
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/expressions/Expression.tsx:31]: 💬 This one is a single Expression card, supporting every type of expression.

File: public/app/features/alerting/unified/components/expressions/AlertCondition.tsx [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+import { GrafanaTheme2 } from '@grafana/data';
+import { Badge, useStyles2 } from '@grafana/ui';
+
+interface AlertConditionProps {
+  enabled?: boolean;
+  onSetCondition: () => void;
+}
+
+export const AlertCondition: FC<AlertConditionProps> = ({ enabled = false, onSetCondition }) => {
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/expressions/AlertCondition.tsx:12]: 💬 A simple component to toggle between "make this the alert condition" and "this is the alert condition".

File: public/app/features/alerting/unified/__snapshots__/PanelAlertTabContent.test.tsx.snap [hunk 17]
```diff
@@ -0,0 +1,117 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`PanelAlertTabContent Will render alerts belonging to panel and a button to create alert from panel queries 1`] = `
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/__snapshots__/PanelAlertTabContent.test.tsx.snap:3]: 💬 I've moved the output of one of the tests in `PanelAlertTabContent` that asserts the data structure of a default alert to a snapshot here as it's easier to work with and update when we change it later (to a reduce + threshold expression).

File: public/app/features/alerting/unified/PanelAlertTabContent.test.tsx [hunk 18]
```diff
@@ -302,58 +301,7 @@ describe('PanelAlertTabContent', () => {
     expect(match).toHaveLength(2);
 
     const defaults = JSON.parse(decodeURIComponent(match![1]));
-    expect(defaults).toEqual({
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/PanelAlertTabContent.test.tsx:305]: 💬 I've moved the output of this test that asserts the data structure of a default alert to a snapshot as it's easier to work with and update when we change it later (to a reduce + threshold expression).

**Comment by jessover9000** [comment]
Cool, this looks good! In the screenshot the expressions are not the same width, shouldn't they have an equally distributed width? Or why is the "Reduce" expression wider than the other one? 🤔 Also, do you have plans how to show that one expression is based on the result of another expression vs based on the value from the query?

**Comment by soniaAguilarPeiron** [comment]
Awesome!! 🙌 This looks really simple and more understandable !! 🎉

One question: it seems we lose some feedback if we have an invalid format ? For example, I removed the expression and I don't see the Preview alerts showing the Error state as we have before:

Screenshot 2022-09-09 at 11 30 27

This is what happen:


https://user-images.githubusercontent.com/33540275/189321452-fd5d28fa-22e1-4c2e-86d3-39f486237cf9.mp4

**Comment by gillesdemey** [comment]
>  In the screenshot the expressions are not the same width, shouldn't they have an equally distributed width? Or why is the "Reduce" expression wider than the other one?

I'll try to dig in to that, it's much wider than it needs to be but I don't think I can make them all the same width give that each expression might render a different set of form elements.

> do you have plans how to show that one expression is based on the result of another expression vs based on the value from the query?

The complication here is that the `math` expression is sort of hard to parse as it could have references to one or more expressions (that may be a mix of alert and expression queries).

File: pkg/services/publicdashboards/api/api_test.go [hunk 19]
```diff
@@ -363,7 +363,7 @@ func TestApiSavePublicDashboardConfig(t *testing.T) {
 	}
 }
 
-// `/public/dashboards/:uid/query`` endpoint test
+// `/public/dashboards/:uid/query“ endpoint test
```
Comment (gillesdemey) [inline pkg/services/publicdashboards/api/api_test.go:366]: I'm not exactly sure what tooling made these changes but I will revert these

**Comment by gillesdemey** [comment]
@soniaAguilarPeiron  

> One question: it seems we lose some feedback if we have an invalid format ? 
Yea, currently I'm not using the preview endpoint at all and added the logic in the front-end to determine if an alert instance will be created.

I should probably still call it to determine if an error would occur so we don't have to add that logic to the front-end.

I'll see if I can figure out how to combine the new redesign with the response that comes from the preview endpoint – perhaps by matching by unique set of labels 🤔

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+  } = useFormContext<RuleFormValues>();
+  const [panelData, setPanelData] = useState<Record<string, PanelData>>({});
+
+  const initialState = {
+    queries: getValues('queries'),
+    panelData: {},
+  };
+  const [{ queries }, dispatch] = useReducer(queriesAndExpressionsReducer, initialState);
+
+  const type = watch('type');
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx:55]: You can try `[type, condition, dataSourceName = watch(['type', 'condition', 'dataSourceName]` to save a few lines

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts (hunk 7 above)
Comment (konrad147) [reply public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts:36]: Nice idea! I like the Immer's approach, I find it less error-prone

Comment (konrad147) [reply public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts:36]: Great job with the reducer for queries/expressions handling!

File: public/app/features/alerting/unified/components/expressions/Expression.tsx (hunk 14 above)
Comment (gillesdemey) [reply public/app/features/alerting/unified/components/expressions/Expression.tsx:229]: `justify-content: space-between` behaves a bit differently, it's essentially the same of only two child components are involved and the `<Flex />` component is in the center, but if the `<Flex />` component is in another place (such as `<div /><div /><Flex />`) it's not going to work with just `justify-content` :(

**Comment by gillesdemey** [comment]
I've added support for visualizing time series data and a warning badge when the user tries to use it as an alert condition (we do not support this).

image

File: public/app/features/alerting/unified/components/expressions/Expression.tsx [hunk 21]
```diff
@@ -0,0 +1,327 @@
...
+
+  const groupedByState = {
+    [PromAlertingRuleState.Firing]: series.filter((serie) => getSeriesValue(serie) >= 1),
+    [PromAlertingRuleState.Inactive]: series.filter((serie) => getSeriesValue(serie) < 1),
+  };
+
+  const renderExpressionType = useCallback(
+    (query: ExpressionQuery) => {
+      // these are the refs we can choose from that don't include the current one
+      const availableRefIds = queries!
```
[the remark this answers is not in the document]
Comment (gillesdemey) [reply public/app/features/alerting/unified/components/expressions/Expression.tsx:64]: Good catch! Probably a left-over from a small refactor I forgot to clean up

File: public/app/features/alerting/unified/components/expressions/Expression.tsx [hunk 22]
```diff
@@ -0,0 +1,327 @@
...
+  refId: string;
+  queryType: ExpressionQueryType;
+  onUpdateRefId: (refId: string) => void;
+  onRemoveExpression: () => void;
+  onUpdateExpressionType: (type: ExpressionQueryType) => void;
+}
+
+const Header: FC<HeaderProps> = ({ refId, queryType, onUpdateRefId, onUpdateExpressionType, onRemoveExpression }) => {
+  const styles = useStyles2(getStyles);
+  const [editMode, setEditMode] = useState<'refId' | 'expressionType' | false>(false);
```
[the remark this answers is not in the document]
Comment (gillesdemey) [reply public/app/features/alerting/unified/components/expressions/Expression.tsx:166]: Right I'll try to clarify and add it as a comment to the code. There are 3 edit modes: 

1. editing the refId (ie. A -> B)
2. editing the type of the expression (ie. Reduce -> Math)
3. `false` means we're not editing either of those

**Comment by soniaAguilarPeiron** [comment]
It seems like the `Run queries` button is not working?

https://user-images.githubusercontent.com/33540275/191712615-b94cff8a-107e-4ba4-82ce-dd281ad85248.mp4

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+    runner.current.cancel();
+  }, []);
+
+  const runQueries = useCallback(() => {
+    runner.current.run(queries);
+  }, [queries]);
+
+  // whenever we update the queries we have to update the form too
+  useEffect(() => {
+    setValue('queries', queries, { shouldValidate: false });
```
Comment (VikaCep) [inline public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx:73]: I remember you mentioned this as an important sync step. Is there already/are you planning to add a test to cover this?

**Review by VikaCep — COMMENTED** [review]
Really nice work!

**Comment by gillesdemey** [comment]
I've hooked up the "run queries" button but it's a bit buggy – will create a follow-up PR to fix that.

**Comment by gillesdemey** [comment]
@soniaAguilarPeiron 

> One question: it seems we lose some feedback if we have an invalid format ? For example, I removed the expression and I don't see the Preview alerts showing the Error state as we have before:

Yea, this is the one thing I haven't added back in here – I'll add that in a follow-up PR and see if we can prevent those rules from being saved in the first place (currently you can save those and they'll just fail when we evaluate them).

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx (hunk 9 above)
Comment (konrad147) [reply public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx:103]: For queries, we display a validation message saying `Query name already exists` but for expressions, something weird happens:
Kapture 2022-09-29 at 09 58 57

File: public/app/features/alerting/unified/utils/rule-form.ts (hunk 6 above)
Comment (konrad147) [reply public/app/features/alerting/unified/utils/rule-form.ts:220]: I wonder if we should also change the default expression that gets added when you click `+ New expression`? Currently, we add the classic condition expression

Comment (gillesdemey) [reply public/app/features/alerting/unified/utils/rule-form.ts:220]: Good question, I think it would make sense – perhaps we should use the  `reduce` expression as the default, or the `threshold`? What do you think?

File: public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx (hunk 9 above)
Comment (gillesdemey) [reply public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/QueryAndExpressionsStep.tsx:103]: Ah right, this happens because I've forgotten to add `type=button` to the `remove expression` button which makes it think it should be activated when pressing "enter" – really wish this was the default type for buttons 😢

**Review by konrad147 — APPROVED** [review]
Great job! 🎉
